File: runners/dotnet/StepDefinitions/StreamingSteps.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using FluentAssertions;
using TechTalk.SpecFlow;

namespace OpenFga.Sdk.Conformance.StepDefinitions
{
    [Binding]
    public class StreamingSteps
    {
        private readonly TestContext _testContext;

        public StreamingSteps(TestContext testContext)
        {
            _testContext = testContext;
        }

        [When(@"I call ReadChanges with no parameters")]
        public async Task WhenICallReadChangesWithNoParameters()
        {
            if (_testContext.Client == null)
            {
                throw new InvalidOperationException("Client not configured");
            }

            await _testContext.ExecuteApiCallAsync(async () =>
            {
                // ReadChanges typically returns an async enumerable or stream
                // Implementation depends on the .NET SDK structure
                var readChangesRequest = new ReadChangesRequest();
                return await _testContext.Client.ReadChanges(readChangesRequest);
            });
        }

        [When(@"I call ReadChanges with:")]
        public async Task WhenICallReadChangesWithParameters(Table table)
        {
            if (_testContext.Client == null)
            {
                throw new InvalidOperationException("Client not configured");
            }

            var readChangesRequest = new ReadChangesRequest();

            foreach (var row in table.Rows)
            {
                var key = row.Keys.First();
                var value = row[key];

                switch (key.ToLowerInvariant())
                {
                    case "type":
                        readChangesRequest.Type = value;
                        break;
                    case "pagesize":
                        if (int.TryParse(value, out int pageSize))
                        {
                            readChangesRequest.PageSize = pageSize;
                        }
                        break;
                    case "continuationtoken":
                        readChangesRequest.ContinuationToken = value;
                        break;
                    case "from":
                        // TODO: Add timestamp support when .NET SDK supports from parameter
                        _testContext.SavedData["from_timestamp"] = value;
                        break;
                    case "to":
                        // TODO: Add timestamp support when .NET SDK supports to parameter
                        _testContext.SavedData["to_timestamp"] = value;
                        break;
                }
            }

            await _testContext.ExecuteApiCallAsync(async () =>
            {
                return await _testContext.Client.ReadChanges(readChangesRequest);
            });
        }

        [When(@"I call ListObjects with:")]
        public async Task WhenICallListObjectsWithParameters(Table table)
        {
            if (_testContext.Client == null)
            {
                throw new InvalidOperationException("Client not configured");
            }

            var listObjectsRequest = new ListObjectsRequest();

            foreach (var row in table.Rows)
            {
                var key = row.Keys.First();
                var value = row[key];

                switch (key.ToLowerInvariant())
                {
                    case "type":
                        listObjectsRequest.Type = value;
                        break;
                    case "relation":
                        listObjectsRequest.Relation = value;
                        break;
                    case "user":
                        listObjectsRequest.User = value;
                        break;
                    case "pagesize":
                        if (int.TryParse(value, out int pageSize))
                        {
                            listObjectsRequest.PageSize = pageSize;
                        }
                        break;
                    case "continuationtoken":
                        listObjectsRequest.ContinuationToken = value;
                        break;
                }
            }

            await _testContext.ExecuteApiCallAsync(async () =>
            {
                return await _testContext.Client.ListObjects(listObjectsRequest);
            });
        }

        [Then(@"the streaming response should contain (\d+) items")]
        public void ThenTheStreamingResponseShouldContainItems(int expectedCount)
        {
            _testContext.LastError.Should().BeNull();
            _testContext.LastResponse.Should().NotBeNull();

            // Implementation depends on the response structure
            // This is a placeholder for streaming response validation
            var actualCount = GetStreamingItemCount(_testContext.LastResponse);
            actualCount.Should().Be(expectedCount);
        }

        [Then(@"the streaming response should contain at least (\d+) items")]
        public void ThenTheStreamingResponseShouldContainAtLeastItems(int minCount)
        {
            _testContext.LastError.Should().BeNull();
            _testContext.LastResponse.Should().NotBeNull();

            var actualCount = GetStreamingItemCount(_testContext.LastResponse);
            actualCount.Should().BeGreaterOrEqualTo(minCount);
        }

        [Then(@"the streaming response should have continuation token")]
        public void ThenTheStreamingResponseShouldHaveContinuationToken()
        {
            _testContext.LastError.Should().BeNull();
            _testContext.LastResponse.Should().NotBeNull();

            var continuationToken = GetContinuationToken(_testContext.LastResponse);
            continuationToken.Should().NotBeNullOrEmpty();
        }

        [Then(@"the streaming response should not have continuation token")]
        public void ThenTheStreamingResponseShouldNotHaveContinuationToken()
        {
            _testContext.LastError.Should().BeNull();
            _testContext.LastResponse.Should().NotBeNull();

            var continuationToken = GetContinuationToken(_testContext.LastResponse);
            continuationToken.Should().BeNullOrEmpty();
        }

        [Then(@"each streaming item should have required fields")]
        public void ThenEachStreamingItemShouldHaveRequiredFields()
        {
            _testContext.LastError.Should().BeNull();
            _testContext.LastResponse.Should().NotBeNull();

            // Validate that each item in the streaming response has required fields
            ValidateStreamingItems(_testContext.LastResponse);
        }

        private int GetStreamingItemCount(object response)
        {
            // Implementation depends on the .NET SDK response structure
            // This is a placeholder - actual implementation would extract count from response
            if (response == null) return 0;
            
            // Example: if response has a Changes or Objects property
            var responseType = response.GetType();
            var changesProperty = responseType.GetProperty("Changes");
            var objectsProperty = responseType.GetProperty("Objects");
            
            if (changesProperty != null)
            {
                var changes = changesProperty.GetValue(response) as System.Collections.IEnumerable;
                return changes?.Cast<object>().Count() ?? 0;
            }
            
            if (objectsProperty != null)
            {
                var objects = objectsProperty.GetValue(response) as System.Collections.IEnumerable;
                return objects?.Cast<object>().Count() ?? 0;
            }
            
            return 0;
        }

        private string GetContinuationToken(object response)
        {
            // Implementation depends on the .NET SDK response structure
            if (response == null) return null;
            
            var responseType = response.GetType();
            var tokenProperty = responseType.GetProperty("ContinuationToken");
            
            return tokenProperty?.GetValue(response) as string;
        }

        // ReadChanges time-based filtering step definitions
        [When(@"I call ReadChanges with from timestamp ""([^""]*)""")]
        public async Task WhenICallReadChangesWithFromTimestamp(string timestamp)
        {
            if (_testContext.Client == null)
            {
                throw new InvalidOperationException("Client not configured");
            }

            // Store timestamp for validation
            _testContext.SavedData["from_timestamp"] = timestamp;

            var readChangesRequest = new ReadChangesRequest();
            // TODO: Add timestamp support when .NET SDK supports from parameter
            // readChangesRequest.From = timestamp;

            await _testContext.ExecuteApiCallAsync(async () =>
            {
                return await _testContext.Client.ReadChanges(readChangesRequest);
            });
        }

        [When(@"I call ReadChanges with to timestamp ""([^""]*)""")]
        public async Task WhenICallReadChangesWithToTimestamp(string timestamp)
        {
            if (_testContext.Client == null)
            {
                throw new InvalidOperationException("Client not configured");
            }

            // Store timestamp for validation
            _testContext.SavedData["to_timestamp"] = timestamp;

            var readChangesRequest = new ReadChangesRequest();
            // TODO: Add timestamp support when .NET SDK supports to parameter
            // readChangesRequest.To = timestamp;

            await _testContext.ExecuteApiCallAsync(async () =>
            {
                return await _testContext.Client.ReadChanges(readChangesRequest);
            });
        }

        [When(@"I call ReadChanges with time range:")]
        public async Task WhenICallReadChangesWithTimeRange(Table table)
        {
            if (_testContext.Client == null)
            {
                throw new InvalidOperationException("Client not configured");
            }

            var readChangesRequest = new ReadChangesRequest();

            foreach (var row in table.Rows)
            {
                var key = row.Keys.First();
                var value = row[key];

                switch (key.ToLowerInvariant())
                {
                    case "from":
                        _testContext.SavedData["from_timestamp"] = value;
                        // TODO: Add timestamp support when .NET SDK supports from parameter
                        // readChangesRequest.From = value;
                        break;
                    case "to":
                        _testContext.SavedData["to_timestamp"] = value;
                        // TODO: Add timestamp support when .NET SDK supports to parameter
                        // readChangesRequest.To = value;
                        break;
                }
            }

            await _testContext.ExecuteApiCallAsync(async () =>
            {
                return await _testContext.Client.ReadChanges(readChangesRequest);
            });
        }

        private void ValidateStreamingItems(object response)
        {
            // Implementation depends on the .NET SDK response structure
            // This would validate that each item has required fields like tupleKey, timestamp, etc.
            if (response == null)
            {
                throw new InvalidOperationException("Response is null");
            }
            
            // Placeholder validation - actual implementation would check item structure
            var responseType = response.GetType();
            responseType.Should().NotBeNull("Response should have a valid type");
        }
    }
}

File: runners/java/src/test/java/dev/openfga/sdk/conformance/steps/StreamingSteps.java
package test.java.dev.openfga.sdk.conformance.steps;

import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import io.cucumber.datatable.DataTable;
import org.assertj.core.api.Assertions;
import dev.openfga.sdk.conformance.support.TestContext;

import java.util.List;
import java.util.Map;

public class StreamingSteps {
    
    private final TestContext testContext;

    public StreamingSteps(TestContext testContext) {
        this.testContext = testContext;
    }

    @When("I call ReadChanges with no parameters")
    public void iCallReadChangesWithNoParameters() {
        if (testContext.getClient() == null) {
            throw new RuntimeException("Client not configured");
        }

        testContext.executeApiCall(() -> {
            // ReadChanges implementation for Java SDK
            // Implementation depends on the Java SDK structure
            ReadChangesRequest request = new ReadChangesRequest();
            return testContext.getClient().readChanges(request).get();
        });
    }

    @When("I call ReadChanges with:")
    public void iCallReadChangesWithParameters(DataTable dataTable) {
        if (testContext.getClient() == null) {
            throw new RuntimeException("Client not configured");
        }

        ReadChangesRequest request = new ReadChangesRequest();
        
        List<Map<String, String>> rows = dataTable.asMaps(String.class, String.class);
        for (Map<String, String> row : rows) {
            String key = row.keySet().iterator().next();
            String value = row.get(key);
            
            switch (key.toLowerCase()) {
                case "type":
                    request.type(value);
                    break;
                case "pagesize":
                    try {
                        request.pageSize(Integer.parseInt(value));
                    } catch (NumberFormatException e) {
                        // Handle invalid page size
                    }
                    break;
                case "continuationtoken":
                    request.continuationToken(value);
                    break;
                case "from":
                    // TODO: Add timestamp support when Java SDK supports from parameter
                    testContext.getSavedData().put("from_timestamp", value);
                    break;
                case "to":
                    // TODO: Add timestamp support when Java SDK supports to parameter
                    testContext.getSavedData().put("to_timestamp", value);
                    break;
            }
        }

        testContext.executeApiCall(() -> {
            return testContext.getClient().readChanges(request).get();
        });
    }

    @When("I call ListObjects with:")
    public void iCallListObjectsWithParameters(DataTable dataTable) {
        if (testContext.getClient() == null) {
            throw new RuntimeException("Client not configured");
        }

        ListObjectsRequest request = new ListObjectsRequest();
        
        List<Map<String, String>> rows = dataTable.asMaps(String.class, String.class);
        for (Map<String, String> row : rows) {
            String key = row.keySet().iterator().next();
            String value = row.get(key);
            
            switch (key.toLowerCase()) {
                case "type":
                    request.type(value);
                    break;
                case "relation":
                    request.relation(value);
                    break;
                case "user":
                    request.user(value);
                    break;
                case "pagesize":
                    try {
                        request.pageSize(Integer.parseInt(value));
                    } catch (NumberFormatException e) {
                        // Handle invalid page size
                    }
                    break;
                case "continuationtoken":
                    request.continuationToken(value);
                    break;
            }
        }

        testContext.executeApiCall(() -> {
            return testContext.getClient().listObjects(request).get();
        });
    }

    @Then("the streaming response should contain {int} items")
    public void theStreamingResponseShouldContainItems(int expectedCount) {
        Assertions.assertThat(testContext.getLastError()).isNull();
        Assertions.assertThat(testContext.getLastResponse()).isNotNull();

        int actualCount = getStreamingItemCount(testContext.getLastResponse());
        Assertions.assertThat(actualCount).isEqualTo(expectedCount);
    }

    @Then("the streaming response should contain at least {int} items")
    public void theStreamingResponseShouldContainAtLeastItems(int minCount) {
        Assertions.assertThat(testContext.getLastError()).isNull();
        Assertions.assertThat(testContext.getLastResponse()).isNotNull();

        int actualCount = getStreamingItemCount(testContext.getLastResponse());
        Assertions.assertThat(actualCount).isGreaterThanOrEqualTo(minCount);
    }

    @Then("the streaming response should have continuation token")
    public void theStreamingResponseShouldHaveContinuationToken() {
        Assertions.assertThat(testContext.getLastError()).isNull();
        Assertions.assertThat(testContext.getLastResponse()).isNotNull();

        String continuationToken = getContinuationToken(testContext.getLastResponse());
        Assertions.assertThat(continuationToken).isNotEmpty();
    }

    @Then("the streaming response should not have continuation token")
    public void theStreamingResponseShouldNotHaveContinuationToken() {
        Assertions.assertThat(testContext.getLastError()).isNull();
        Assertions.assertThat(testContext.getLastResponse()).isNotNull();

        String continuationToken = getContinuationToken(testContext.getLastResponse());
        Assertions.assertThat(continuationToken).isNullOrEmpty();
    }

    @Then("each streaming item should have required fields")
    public void eachStreamingItemShouldHaveRequiredFields() {
        Assertions.assertThat(testContext.getLastError()).isNull();
        Assertions.assertThat(testContext.getLastResponse()).isNotNull();

        validateStreamingItems(testContext.getLastResponse());
    }

    private int getStreamingItemCount(Object response) {
        // Implementation depends on the Java SDK response structure
        if (response == null) return 0;
        
        // Example: if response has getChanges() or getObjects() method
        try {
            var responseClass = response.getClass();
            
            // Try to get changes
            try {
                var getChangesMethod = responseClass.getMethod("getChanges");
                var changes = (List<?>) getChangesMethod.invoke(response);
                return changes != null ? changes.size() : 0;
            } catch (Exception e) {
                // Method doesn't exist, try objects
            }
            
            // Try to get objects
            try {
                var getObjectsMethod = responseClass.getMethod("getObjects");
                var objects = (List<?>) getObjectsMethod.invoke(response);
                return objects != null ? objects.size() : 0;
            } catch (Exception e) {
                // Method doesn't exist
            }
            
        } catch (Exception e) {
            // Handle reflection errors
        }
        
        return 0;
    }

    private String getContinuationToken(Object response) {
        // Implementation depends on the Java SDK response structure
        if (response == null) return null;
        
        try {
            var responseClass = response.getClass();
            var getContinuationTokenMethod = responseClass.getMethod("getContinuationToken");
            return (String) getContinuationTokenMethod.invoke(response);
        } catch (Exception e) {
            // Method doesn't exist or other error
            return null;
        }
    }

    private void validateStreamingItems(Object response) {
        // Implementation depends on the Java SDK response structure
        if (response == null) {
            throw new RuntimeException("Response is null");
        }
        
        // Placeholder validation - actual implementation would check item structure
        Assertions.assertThat(response).isNotNull();
        
        // Additional validation could check that each item has required fields
        // like tupleKey, timestamp, etc.
    }

    // ReadChanges time-based filtering step definitions
    @When("I call ReadChanges with from timestamp {string}")
    public void iCallReadChangesWithFromTimestamp(String timestamp) throws Exception {
        if (testContext.getClient() == null) {
            throw new IllegalStateException("Client not configured");
        }

        // Store timestamp for validation
        testContext.getSavedData().put("from_timestamp", timestamp);

        testContext.executeApiCall(() -> {
            ReadChangesRequest request = new ReadChangesRequest();
            // TODO: Add timestamp support when Java SDK supports from parameter
            // request.from(timestamp);
            return testContext.getClient().readChanges(request).get();
        });
    }

    @When("I call ReadChanges with to timestamp {string}")
    public void iCallReadChangesWithToTimestamp(String timestamp) throws Exception {
        if (testContext.getClient() == null) {
            throw new IllegalStateException("Client not configured");
        }

        // Store timestamp for validation
        testContext.getSavedData().put("to_timestamp", timestamp);

        testContext.executeApiCall(() -> {
            ReadChangesRequest request = new ReadChangesRequest();
            // TODO: Add timestamp support when Java SDK supports to parameter
            // request.to(timestamp);
            return testContext.getClient().readChanges(request).get();
        });
    }

    @When("I call ReadChanges with time range:")
    public void iCallReadChangesWithTimeRange(DataTable dataTable) throws Exception {
        if (testContext.getClient() == null) {
            throw new IllegalStateException("Client not configured");
        }

        ReadChangesRequest request = new ReadChangesRequest();
        
        List<Map<String, String>> rows = dataTable.asMaps(String.class, String.class);
        for (Map<String, String> row : rows) {
            String key = row.keySet().iterator().next();
            String value = row.get(key);
            
            switch (key.toLowerCase()) {
                case "from":
                    testContext.getSavedData().put("from_timestamp", value);
                    // TODO: Add timestamp support when Java SDK supports from parameter
                    // request.from(value);
                    break;
                case "to":
                    testContext.getSavedData().put("to_timestamp", value);
                    // TODO: Add timestamp support when Java SDK supports to parameter
                    // request.to(value);
                    break;
            }
        }

        testContext.executeApiCall(() -> {
            return testContext.getClient().readChanges(request).get();
        });
    }
}
